src/passes/ConstantFieldPropagation.cpp

```diff
@@ -113,13 +113,82 @@ struct FunctionOptimizer : public WalkerPass<PostWalker<FunctionOptimizer>> {
     : propagatedInfos(propagatedInfos), subTypes(subTypes),
       rawNewInfos(rawNewInfos), refTest(refTest) {}
 
-  void visitStructGet(StructGet* curr) {
+  template<typename T> std::optional<HeapType> getRelevantHeapType(T* curr) {
     auto type = curr->ref->type;
     if (type == Type::unreachable) {
-      return;
+      return std::nullopt;
     }
     auto heapType = type.getHeapType();
     if (!heapType.isStruct()) {
+      return std::nullopt;
+    }
+    return heapType;
+  }
+
+  PossibleConstantValues getInfo(HeapType type, Index index) {
+    if (auto it = propagatedInfos.find(type); it != propagatedInfos.end()) {
+      // There is information on this type, fetch it.
+      return it->second[index];
+    }
+    return PossibleConstantValues{};
+  }
+
+  template<typename T> Expression* makeDroppedRef(T* curr) {
```

Comment:
```suggestion
    // Returns a block containing the dropped input.
    template<typename T> Block* getDroppingBlock(T* curr) {
```
This calls blockify, so we might as well keep the type information that it returns. We can then also just call `block->list.push_back` in the places below.